direct/io-kafka/src/main/java/cz/o2/proxima/direct/kafka/KafkaLogReader.java

```diff
@@ -479,6 +456,55 @@ private void submitConsumerWithObserver(
     latch.await();
   }
 
+  private ConsumerRecords<Object, Object> seekToNewOffsetsIfNeeded(
+      @Nullable Collection<Offset> offsets,
+      ElementConsumer<Object, Object> consumer,
+      List<TopicOffset> seekOffsets,
+      AtomicReference<PartitionedWatermarkEstimator> watermarkEstimator,
+      KafkaConsumer<Object, Object> kafka,
+      ConsumerRecords<Object, Object> poll) {
+
+    synchronized (seekOffsets) {
+      if (!seekOffsets.isEmpty()) {
+        Utils.seekToOffsets(topic, offsets, kafka);
+        consumer.onAssign(
+            kafka,
+            kafka
+                .assignment()
+                .stream()
+                .map(
+                    tp ->
+                        new TopicOffset(
+                            tp.partition(),
+                            kafka.position(tp),
+                            watermarkEstimator.get().getWatermark()))
```

Comment:
Null check not needed here?

Reply:
No, because when the code arrives here, the watermark estimator is already initialized (by listener).